### ISSUE TITLE
BUILD FAILURE on frontend-maven-plugin:0.0.25:npm

### ISSUE THREAD

**Issue by canoine** [Athou/commafeed#781]
Hello,

I'm trying to install commafeed on a new server. But unsuccessfully :

```
[INFO] --- frontend-maven-plugin:0.0.25:npm (npm install) @ commafeed ---
[INFO] Running 'npm install --loglevel info --color=false' in /var/local/commafeed/src/commafeed
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 10.875 s
[INFO] Finished at: 2016-02-12T19:18:26+01:00
[INFO] Final Memory: 45M/384M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:0.0.25:npm (npm install) on project commafeed: Failed to run task: 'npm install --loglevel info --color=false' failed. (error code 139) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.github.eirslett:frontend-maven-plugin:0.0.25:npm (npm install) on project commafeed: Failed to run task
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:862)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:286)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:197)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoFailureException: Failed to run task
        at com.github.eirslett.maven.plugins.frontend.mojo.AbstractFrontendMojo.execute(AbstractFrontendMojo.java:66)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 20 more
Caused by: com.github.eirslett.maven.plugins.frontend.lib.TaskRunnerException: 'npm install --loglevel info --color=false' failed. (error code 139)
        at com.github.eirslett.maven.plugins.frontend.lib.NodeTaskExecutor.execute(NodeTaskExecutor.java:59)
        at com.github.eirslett.maven.plugins.frontend.mojo.NpmMojo.execute(NpmMojo.java:50)
        at com.github.eirslett.maven.plugins.frontend.mojo.AbstractFrontendMojo.execute(AbstractFrontendMojo.java:64)
        ... 22 more
[ERROR]
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```
As building the same Commafeed version works on my "production server", which runs roughly the same softwares, on the same OS, I'm looking for the difference explaining the failure on one, and the sucess on the other. But I can't find any more precise hint than "Failed to run task", which means approximately nothing for me.
How can I find some "sysadmin-friendly" logs ?

**Comment by Athou** [comment]
I don't know why it does not build on your system, you could try to change the loglevel of npm to verbose (line 184 of pom.xml)

**Comment by canoine** [comment]
I found this line, and even set loglevel to debug. I also launched mvn with the -X flag. That produces lots of logs, which may be eventually useful, if you want them.
Anyway, I found, and solved, the problem, following other paths. So I think providing the logs are useless, now.

When I said the servers are roughly identical, I forgot a small little insignificant difference... Hem... Called grsecurity.
If I understand well, node/npm is a sh script, which calls the node/node binary. But, when called directly, node tries a "denied RWX mmap of \<anonymous mapping\>" (I don't know exactly what that means, I'll try to have a look on it when I can) for grsecurity, so it segfaults. strace said : 
```
(...)
mmap(0x34975fbb000, 4096, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = -1 EPERM (Operation not permitted)
munmap(0x34975fbb000, 4096)             = 0                                       
mmap(0x12781ad8000, 4096, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = -1 EPERM (Operation not permitted)
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x20} ---
+++ killed by SIGSEGV +++
Segmentation fault
```
Disabling PaX MPROTECT flag for the node binary unblocks the situation. For the record : 
```
# paxctl-ng -c node/node
# paxctl-ng -m node/node
```
I don't know if I should open an issue on the node project's repo (I suppose it's nodejs/node, on Github, no ?), or if it's a compilation-related issue. When building Commafeed, does frontend-maven-plugin download node/node as a binary, or as source, which is compiled on the final server ?

**Comment by josecelaya** [comment]
I have the same problem trying to build on a Max. Is there any resolution to this?
cheers
J

**Comment by canoine** [comment]
What is a Max ?
If you're trying to build Commafeed on a hardened system, you need to deactivate some security enforcements for the commafeed/node/node binary. I found how to do with grsecurity (the two "paxctl-ng" commands above solve the problem), but I think you can adapt the same principle, with different commands, to other hardening softs.

**Comment by Athou** [comment]
Note that you can build commafeed on any machine, the resulting jar file is cross-platform and all dependencies are embedded.

I just released 2.3.0 too, a binary jar is available on the release page.

**Comment by canoine** [comment]
As I the main problem is solved, and all works well since, I think this issue can be closed.
